metagraph/src/graph/representation/canonical_dbg.hpp

```diff
@@ -141,6 +142,10 @@ class CanonicalDBG : public DBGWrapper<DeBruijnGraph> {
 
     std::array<size_t, 256> alphabet_encoder_;
 
+    // a thread-safe 0-size NodeFirstCache to use as a fallback if another NodeFirstCache
+    // extension is not available
+    mutable std::unique_ptr<NodeFirstCache> fallback_cache_;
```

Comment:
Why not create a new cache for each CanonicalDBG wrapper?

Reply:
In the ideal case (see align.cpp), one CanonicalDBG wrapper and one NodeFirstCache exists for each thread. However, if a CanonicalDBG is being used by multiple threads, then we don't create a NodeFirstCache for it, so this acts as a fallback to do the same operations cache-free.

Reply:
You could create multiple CanonicalDBG wrappers, one for each thread, and each of the wrappers would have its cache in it.

Reply:
not really. the caches are accessed very often, so having a shared cache for each thread creates too much contention

Reply:
Hm, I'd think it's way less than fwd/bwd calls that it replaces. Does it really affect the performance that much?

Reply:
there are two things to consider here:
1) when I tested it last time, the CPU usage saturated at around 4000%, so it was pretty severe
2) the cache is most effective locally. 10 threads aligning different read batches are very unlikely to be accessing the same part of the graph within a reasonable timeframe of each other, so it doesn't make much sense to have a shared cache

Reply:
If you create a cache for each thread, every new query will start with an empty cache. Though, the queries might be actually similar... What do you think?

Reply:
I tried this out on a machine with 128 threads. It looks like saturates at 16 threads when having a shared cache, but we don't know when it saturates for the separate caches (I assume that it slows down at 128 threads because of hyperthreading)

```
Nthreads	Separate caches							Shared cache
128		Elapsed (wall clock) time (h:mm:ss or m:ss): 0:41.13		Elapsed (wall clock) time (h:mm:ss or m:ss): 0:56.94
64		Elapsed (wall clock) time (h:mm:ss or m:ss): 0:37.88		Elapsed (wall clock) time (h:mm:ss or m:ss): 0:56.27
32		Elapsed (wall clock) time (h:mm:ss or m:ss): 0:38.86		Elapsed (wall clock) time (h:mm:ss or m:ss): 0:51.65
16		Elapsed (wall clock) time (h:mm:ss or m:ss): 0:46.97		Elapsed (wall clock) time (h:mm:ss or m:ss): 0:49.63
8		Elapsed (wall clock) time (h:mm:ss or m:ss): 1:16.69		Elapsed (wall clock) time (h:mm:ss or m:ss): 1:11.30
4		Elapsed (wall clock) time (h:mm:ss or m:ss): 2:19.48		Elapsed (wall clock) time (h:mm:ss or m:ss): 2:03.94
2		Elapsed (wall clock) time (h:mm:ss or m:ss): 4:21.65		Elapsed (wall clock) time (h:mm:ss or m:ss): 3:46.17
1		Elapsed (wall clock) time (h:mm:ss or m:ss): 9:05.37		Elapsed (wall clock) time (h:mm:ss or m:ss): 7:37.99
```

Reply:
Hard to say. Would be interesting to compare, single cache vs multiple for 64 threads